File: Data Structure/Stack/LET_no496_Next Greater Element I.py
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        stack, table = [], {}
        for n in nums2:
            while stack and stack[-1] < n:
                table[stack.pop()] = n
            stack.append(n)
        res = []
        for n in nums1:
            res.append(table.get(n, -1))
        return res

File: Data Structure/Stack/LET_no496_Next Greater Element I.cpp
class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
        stack<int> s;
        unordered_map<int, int> table;
        for (int n : nums2) {
            while (s.size() && s.top() < n) {
                table[s.top()] = n;
                s.pop();
            }
            s.push(n);
        }
        
        int n = nums1.size();
        vector<int> res(n, -1);
        for (int i = 0; i < n; i++) {
            auto it = table.find(nums1[i]);
            if (it != table.end())
                res[i] = it->second;
        }
        return res;
    }
};